[Dixit API] create dixit / story / playcard / guess

wallychen0705 opened the pull request:
[POST] /api/dixit
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/story
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/playcard
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/guess

wallychen0705 committed:
[POST] /api/dixit
[POST] /api/dixit/{dixitId}/rounds/{round}/story

wallychen0705 committed:
[POST] /api/dixit
[PUT] /api/dixit/{dixitId}/rounds/{round}/story
[PUT] /api/dixit/{dixitId}/rounds/{round}/playcard
[PUT] /api/dixit/{dixitId}/rounds/{round}/guess

Johnny850807 commented on spring-boot-dixit/src/main/java/tw/wally/dixit/controllers/DixitController.java:
```diff
@@ -0,0 +1,58 @@
+        playCardUseCase.execute(request);
+    }
+
+    @PutMapping("/{dixitId}/rounds/{round}/guess")
```
Not RESTFul (who guesses?)

Johnny850807 commented on spring-boot-dixit/src/main/java/tw/wally/dixit/controllers/DixitController.java:
```diff
@@ -0,0 +1,58 @@
+        tellStoryUseCase.execute(request);
+    }
+
+    @PutMapping("/{dixitId}/rounds/{round}/playcard")
```
Not RESTFul (who plays?)

Johnny850807 reviewed (changes requested):
Reviewed your API design

wallychen0705 committed:
add CI environment variables

wallychen0705 committed:
[POST] /api/dixit
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/story
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/playcard
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/guess

Johnny850807 commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+                .orElseThrow()
+                .filter(playCard -> !playCard.getPlayer().equals(gamer))
+                .map(PlayCard::getCard)
+                .findFirst().orElseThrow();
```
Don't leak the internal structure of Dixit.

Consider adding a new method `getGuessedCard(player:Player): Optional<Card>` in Dixit

Johnny850807 commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+                .orElseThrow()
+                .map(Guess::getGuesser)
+                .collect(toList());
+        assertEqualsIgnoreOrder(expectedGamers, actualGamers);
```
Ditto

Johnny850807 reviewed (changes requested):
Review for a while

wallychen0705 committed:
[POST] /api/dixit
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/story
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/playcard
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/guess

Johnny850807 commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,261 @@
+        assertEquals(expectedStoryteller, actualStoryteller);
+    }
+
+    private void assertEachGamerPlayCard(Collection<Player> expectedGamers) {
```
They are not gamers.

UL

Johnny850807 commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    private void assertEachGuesserGuessStory(Collection<Player> expectedGamers) {
+        var actualGamers = dixitRepository.findDixitById(DIXIT_ID)
+                .map(Dixit::getCurrentGuessersWhoGuessed)
+                .orElseThrow();
+        assertEqualsIgnoreOrder(expectedGamers, actualGamers);
```
They are not gamers.

UL

Johnny850807 reviewed (changes requested):
Rename all of the gamers to players

They are all the players and only players.

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/CreateDixitUseCase.java:
```diff
@@ -0,0 +1,82 @@
+    private final CardRepository cardRepository;
+
+    public void execute(Request request) {
+        var cards = cardRepository.findAll();
+        var game = request.game;
+        var dixit = new Dixit(game.id, game.gameSetting.toVictoryCondition(), cards);
+        dixit.join(game.host.toPlayer());
+        mapToList(game.players, Gamer::toPlayer).forEach(dixit::join);
+        dixit.start();
+        dixit = dixitRepository.save(dixit);
+        mayPublishEvents(dixit);
```
Don't mix the DTO conversion code with the use case flow.

```java
        List<Card> cards = cardRepository.findAll();
        Dixit dixit = dixit(request, cards);

        players(request).forEach(dixit::join);
        dixit.start();

        dixit = dixitRepository.save(dixit);
        mayPublishEvents(dixit);
```

Johnny850807 replied:
@Wally5077 any update?

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/GuessStoryUseCase.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public void execute(Request request) {
+        var dixit = findDixit(request.gameId);
```
Don't `var`

No benefits on this.

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/GuessStoryUseCase.java:
```diff
@@ -0,0 +1,38 @@
+        var dixit = findDixit(request.gameId);
+        validateRound(dixit, request.round);
+        var playerId = request.playerId;
+        var guesser = findFirst(dixit.getCurrentGuessers(), player -> playerId.equals(player.getId()))
+                .orElseThrow(() -> new InvalidGameOperationException(format("Player: %s is not guesser", playerId)));
```
`Player guesser = dixit.getGuesser(playerId);`

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/GuessStoryUseCase.java:
```diff
@@ -0,0 +1,38 @@
+        var playerId = request.playerId;
+        var guesser = findFirst(dixit.getCurrentGuessers(), player -> playerId.equals(player.getId()))
+                .orElseThrow(() -> new InvalidGameOperationException(format("Player: %s is not guesser", playerId)));
+        var playCard = dixit.getPlayCardByCardId(request.cardId);
```
`PlayCard playCard =  dixit.getPlayCard(request.cardId);`

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/PlayCardUseCase.java:
```diff
@@ -0,0 +1,38 @@
+        super(dixitRepository);
+    }
+
+    public void execute(Request request) {
+        var dixit = findDixit(request.gameId);
+        validateRound(dixit, request.round);
+        var playerId = request.playerId;
+        var guesser = findFirst(dixit.getCurrentGuessers(), player -> playerId.equals(player.getId()))
+                .orElseThrow(() -> new InvalidGameOperationException(format("Player: %s is not guesser", playerId)));
+        var card = guesser.playCard(request.cardId);
+        dixit.playCard(guesser, card);
+        dixit = dixitRepository.save(dixit);
+        mayPublishEvents(dixit);
+    }
```
ditto

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/TellStoryUseCase.java:
```diff
@@ -0,0 +1,59 @@
+        super(dixitRepository);
+    }
+
+    public void execute(Request request) {
+        var dixit = findDixit(request.gameId);
+        validateRound(dixit, request.round);
+        validateStoryteller(dixit, request.playerId);
+        var storyteller = dixit.getCurrentStoryteller();
+        var card = storyteller.playCard(request.cardId);
+        dixit.tellStory(request.phrase, storyteller, card);
+        dixit = dixitRepository.save(dixit);
+        mayPublishEvents(dixit);
+    }
```
ditto

Johnny850807 reviewed (changes requested):
Review usecases

wallychen0705 committed:
add two different application.yml into .gitignore

wallychen0705 committed:
[POST] /api/dixit
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/story
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/playcard
[PUT] /api/dixit/{dixitId}/rounds/{round}/players/{playerId}/guess

zseqw123 reviewed (approved):
LGTM

wallychen0705 committed:
fix pr

Johnny850807 commented on dixit/src/main/java/tw/wally/dixit/usecases/CreateDixitUseCase.java:
```diff
@@ -0,0 +1,92 @@
+        var cards = cardRepository.findAll();
+        var game = request.game;
+        Dixit dixit = new Dixit(game.id, game.gameSetting.toVictoryCondition(), cards);
+        dixit.join(game.host.toPlayer());
```
should be returned from `players`

Johnny850807 reviewed (changes requested):
Review done

wallychen0705 committed:
fix pr2

Chaoyuuu commented on dixit/src/test/java/tw/wally/dixit/model/DixitTest.java:
```diff
@@ -0,0 +1,194 @@
+
+        assertThrows(InvalidGameOperationException.class, () -> joinGame(players));
+    }
+
```
Why did you test these again? It is the same as `GameTest`.

wallychen0705 replied:
GameTest should be deleted.

Chaoyuuu reviewed (changes requested):
This PR is as huge as WATERBALL!

wallychen0705 committed:
fix pr3

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+            return new FakeCardRepository();
+        }
+    }
+
+    @Test
+    public void WhenCreateDixitWithFourGamers_ThenShouldSuccess() throws Exception {
+        createDixitWithGamers(4)
+                .andExpect(status().isOk());
+
```
Why did you use `gamers` instead of `players`? What is the difference?

wallychen0705 replied:
It was called 'gamer' by lobby service, but finally 'gamer' will become 'player' after discussing.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+    public void WhenCreateDixitWithEightGamers_ThenShouldFail() throws Exception {
+        createDixitWithGamers(8)
+                .andExpect(status().isBadRequest());
+    }
```
rename to `...ThenShouldRespondBadRequest`.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    public void GivenDixitStarted_WhenStorytellerTellStory_ThenCurrentRoundShouldHaveStory() throws Exception {
+        Dixit dixit = createDixitWithGamersAndGet(6);
+        assertNull(dixit.getCurrentRound().getStory());
+
+        var storyteller = dixit.getCurrentStoryteller();
+        tellStory(dixit.getNumberOfRounds(), storyteller)
+                .andExpect(status().isOk());
+
+        assertStorytellerTellStory(storyteller);
+    }
```
`assertStorytellerTellStory(storyteller);` doesn't imply the statement `CurrentRoundShouldHaveStory` to me. 🤔

wallychen0705 replied:
I have changed the method name to new one which is 'assertCurrentRoundHasStoryWhichToldByStoryteller'

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+    @Test
+    public void GivenDixitStarted_WhenGuesserTellStory_ThenShouldFail() throws Exception {
+        Dixit dixit = createDixitWithGamersAndGet(5);
+        assertNull(dixit.getCurrentRound().getStory());
```
`...ThenShouldRespondBadRequest`.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+    @Test
+    public void GivenDixitStartedAndFirstRoundStarted_WhenStorytellerTellStoryAtSecondRound_ThenShouldFail() throws Exception {
+        Dixit dixit = createDixitWithGamersAndGet(6);
+        assertEquals(1, dixit.getNumberOfRounds());
```
`GivenDixitStartedInFirstRound_WhenStorytellerTellsStoryInSecondRound_ThenShouldRespondBadRequest`

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+    @Test
+    public void GivenStoryToldAtFirstRound_WhenOneGuesserPlayCardAtSecondRound_ThenShouldFail() throws Exception {
+        Dixit dixit = givenStoryToldAndGetDixit();
+        assertEquals(1, dixit.getNumberOfRounds());
```
rename to `GivenStoryToldInFirstRound_WhenOneGuesserPlaysCardInSecondRound_ThenShouldRespondBadRequest`.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+        eachGuesserGuessStory(numberOfRounds, guessers);
+
+        assertEachGuesserGuessStory(guessers);
+    }
```
`GivenEachGuesserPlayedCard..` -> `GiveAllGuessersPlayedCard..`
`givenEachGuesserPlayedCardAndGetDixit();` -> `givenAllGuessersPlayedCardAndGetDixit();`

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+        assertEachGuesserGuessStory(guessers);
+    }
+
+    @Test
+    public void GivenEachGuesserPlayedCard_WhenStorytellerGuessStory_ThenShouldFail() throws Exception {
+        Dixit dixit = givenEachGuesserPlayedCardAndGetDixit();
+
+        int numberOfRounds = dixit.getNumberOfRounds();
+        Player storyteller = dixit.getCurrentStoryteller();
+        guessStory(numberOfRounds, storyteller)
+                .andExpect(status().isBadRequest());
+    }
```
Ditto.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    public void GivenEachGuesserPlayedCardAtFirstRound_WhenOneGuesserGuessStoryAtSecondRound_ThenShouldFail() throws Exception {
+        Dixit dixit = givenEachGuesserPlayedCardAndGetDixit();
+        assertEquals(1, dixit.getNumberOfRounds());
+
+        var guesser = dixit.getCurrentGuessers().get(0);
+        guessStory(2, guesser)
+                .andExpect(status().isBadRequest());
+    }
```
rename to `GivenAllGuessersPlayedCardInFirstRound_WhenOneGuesserGuessStoryInSecondRound_ThenShouldRespondBadRequest`

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+    @Test
+    public void GivenDixitStartedAndFirstRoundStarted_WhenStorytellerTellStoryAtSecondRound_ThenShouldFail() throws Exception {
+        Dixit dixit = createDixitWithGamersAndGet(6);
+        assertEquals(1, dixit.getNumberOfRounds());
+
+        Player storyteller = dixit.getCurrentStoryteller();
+        tellStory(2, storyteller)
+                .andExpect(status().isBadRequest());
+    }
```
Is it better to extract `1` to a constant `FIRST_ROUND` and `2` to `SECOUND_RONUD`? 
This is just a suggestion.

Chaoyuuu commented on spring-boot-dixit/src/test/java/tw/wally/dixit/DixitControllerTest.java:
```diff
@@ -0,0 +1,277 @@
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    public void GivenStoryTold_WhenTwoGuessersPlayCard_ThenCurrentRoundShouldHaveTwoPlayCards() throws Exception {
+        Dixit dixit = givenStoryToldAndGetDixit();
+
+        int numberOfRounds = dixit.getNumberOfRounds();
+        var guessers = limit(dixit.getCurrentGuessers(), 2);
+        eachGuesserPlayCard(numberOfRounds, guessers);
+
+        assertEachGuesserPlayCard(guessers);
+    }
```
1. `numberOfRounds` can rename to `currentRound`.
2. `assertEachGuesserPlayCard(guessers);` doesn't imply `ThenCurrentRoundShouldHaveTwoPlayCards`. 
You should rename `assertEachGuesserPlayCard(guessers)` to match the _then_ statement.

Chaoyuuu reviewed (changes requested):
Review Done!

wallychen0705 committed:
fix pr4
